Merge back #21799

#21799 was reverted as it broke `main`. See #21895 for more details.

The following error was reported on multiple merge commits to `main`
```scala
 Wrong number of warnings encountered when compiling tests\warn\i21860.scala
expected: 0, actual: 1

Unexpected warnings:
tests\warn\i21860.scala:11
-> following the diagnostics:
 at 11: Unreachable case
Error:  Test dotty.tools.dotc.CompilationTests.warn failed: java.lang.AssertionError: Warn test failed, but should not, reasons:
Error:  encountered 1 test failure(s):
Error:    - generic failure (see test output), took 18.087 sec
Error:      at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkPass(ParallelTesting.scala:1257)
Error:      at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkWarnings(ParallelTesting.scala:1176)
Error:      at dotty.tools.dotc.CompilationTests.warn(CompilationTests.scala:132)
Error:      ...
```

--- a/compiler/src/dotty/tools/dotc/core/Decorators.scala
+++ b/compiler/src/dotty/tools/dotc/core/Decorators.scala
@@ -292,7 +292,7 @@ object Decorators {
       case _ => String.valueOf(x).nn
 
     /** Returns the simple class name of `x`. */
-    def className: String = x.getClass.getSimpleName.nn
+    def className: String = if x == null then "<null>" else x.getClass.getSimpleName.nn
 
   extension [T](x: T)
     def assertingErrorsReported(using Context): T = {

--- a/compiler/src/dotty/tools/dotc/core/TypeOps.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeOps.scala
@@ -930,7 +930,11 @@ object TypeOps:
       for tp <- mixins.reverseIterator do
         protoTp1 <:< tp
       maximizeType(protoTp1, NoSpan)
-      wildApprox(protoTp1)
+      val inst = wildApprox(protoTp1)
+      if !inst.classSymbol.exists then
+        // E.g. i21790, can't instantiate S#CA as a subtype of O.A, because O.CA isn't accessible
+        NoType
+      else inst
     }
 
     if (protoTp1 <:< tp2) instantiate()

--- a/compiler/src/dotty/tools/dotc/core/TypeOps.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeOps.scala
@@ -931,7 +931,7 @@ object TypeOps:
         protoTp1 <:< tp
       maximizeType(protoTp1, NoSpan)
       val inst = wildApprox(protoTp1)
-      if !inst.classSymbol.exists then
+      if inst.classSymbols.isEmpty then
         // E.g. i21790, can't instantiate S#CA as a subtype of O.A, because O.CA isn't accessible
         NoType
       else inst

--- a/tests/pos/i21790.scala
+++ b/tests/pos/i21790.scala
@@ -1,0 +1,14 @@
+package p
+
+trait S:
+  sealed trait A
+  private class CA() extends A
+
+object O extends S
+
+trait T
+
+class Test:
+  def f(e: T) = e match
+    case _: O.A =>
+    case _      =>

